src/persist-client/src/internal/state_versions.rs

```diff
@@ -0,0 +1,578 @@
+        }
+    }
+
+    /// Updates the state of a shard to a new `current` iff `expected` matches
+    /// `current`.
+    ///
+    /// May be called on uninitialized shards.
+    pub async fn try_compare_and_set_current<K, V, T, D>(
+        &self,
+        cmd_name: &str,
+        shard_metrics: &ShardMetrics,
+        expected: Option<SeqNo>,
+        new_state: &State<K, V, T, D>,
+        diff: &StateDiff<T>,
+    ) -> Result<Result<(), Vec<VersionedData>>, Indeterminate>
+    where
+        K: Debug + Codec,
+        V: Debug + Codec,
+        T: Timestamp + Lattice + Codec64,
+        D: Semigroup + Codec64,
+    {
+        assert_eq!(shard_metrics.shard_id, new_state.shard_id);
+        let path = new_state.shard_id.to_string();
+
+        trace!(
+            "apply_unbatched_cmd {} attempting {}\n  new_state={:?}",
+            cmd_name,
+            new_state.seqno(),
+            new_state
+        );
+        let new = self
+            .metrics
+            .codecs
+            .state_diff
+            .encode(|| VersionedData::from((new_state.seqno(), diff)));
+        assert_eq!(new.seqno, diff.seqno_to);
+
+        let payload_len = new.data.len();
+        let cas_res = retry_determinate(
+            &self.metrics.retries.determinate.apply_unbatched_cmd_cas,
+            || async {
+                self.consensus
+                    .compare_and_set(&path, expected, new.clone())
+                    .await
+            },
+        )
+        .instrument(debug_span!("apply_unbatched_cmd::cas", payload_len))
+        .await
+        .map_err(|err| {
+            debug!("apply_unbatched_cmd {} errored: {}", cmd_name, err);
+            err
+        })?;
+
+        match cas_res {
+            Ok(()) => {
+                trace!(
+                    "apply_unbatched_cmd {} succeeded {}\n  new_state={:?}",
+                    cmd_name,
+                    new_state.seqno(),
+                    new_state
+                );
+
+                shard_metrics.set_since(new_state.since());
+                shard_metrics.set_upper(&new_state.upper());
+                shard_metrics.set_batch_count(new_state.batch_count());
+                shard_metrics.set_update_count(new_state.num_updates());
+                shard_metrics.set_seqnos_held(new_state.seqnos_held());
+                shard_metrics.inc_encoded_diff_size(payload_len);
+                Ok(Ok(()))
+            }
+            Err(live_diffs) => {
+                debug!(
+                    "apply_unbatched_cmd {} {} lost the CaS race, retrying: {} vs {:?}",
+                    new_state.shard_id(),
+                    cmd_name,
+                    new_state.seqno(),
+                    live_diffs.last().map(|x| x.seqno)
+                );
+                Ok(Err(live_diffs))
+            }
+        }
+    }
+
+    /// Fetches the `current` state of the requested shard.
+    ///
+    /// Uses the provided hint, which is a possibly outdated copy of all live
```

Comment:
nit: `hint` comes from before the rename to `all_live_diffs`. Maybe say something like "the provided hint (`all_live_diffs`).

Reply:
done